File: Solutions/42-trapping-rain-water/trapping-rain-water.ts
function trap(height: number[]): number {
    // Total water trapped
    let water = 0;

    // Edge case: if less than 3 bars, no water can be trapped
    if (height.length < 3) return 0;

    // Two pointers starting from left and right ends
    let left = 0;
    let right = height.length - 1;

    // Variables to keep track of the maximum height seen so far from left and right
    let leftMax = 0;
    let rightMax = 0;

    // Loop until the two pointers meet
    while (left <= right) {
        // Decide which pointer to move based on smaller height
        if (height[left] <= height[right]) {
            // If current left bar is smaller or equal, water trapped depends on leftMax
            if (height[left] >= leftMax) {
                // Update leftMax
                leftMax = height[left];
            } else {
                // Water trapped at this position is leftMax - height[left]
                water += leftMax - height[left];
            }
            // Move left pointer to the right
            left++;
        } else {
            // Current right bar is smaller, water trapped depends on rightMax
            if (height[right] >= rightMax) {
                // Update rightMax
                rightMax = height[right];
            } else {
                // Water trapped at this position is rightMax - height[right]
                water += rightMax - height[right];
            }
            // Move right pointer to the left
            right--;
        }
    }

    return water;
}

File: Solutions/42-trapping-rain-water/trapping-rain-water.kt
class Solution {
    fun trap(height: IntArray): Int {

        // We need at least 3 bars to trap any water.
        if (height.size < 3) {
            return 0
        }

        // Two pointers start at the two ends of the array.
        var left = 0
        var right = height.lastIndex

        // Highest bar encountered from the left side.
        var leftMax = 0

        // Highest bar encountered from the right side.
        var rightMax = 0

        // Total amount of trapped water.
        var water = 0

        // Continue until the two pointers meet.
        while (left < right) {

            // We process the side with the smaller
            // current height.
            //
            // Why?
            // If height[left] <= height[right], then the
            // right side already has a wall at least as high
            // as height[left]. Therefore, the amount of water
            // at the left position can be determined entirely
            // by leftMax.
            if (height[left] <= height[right]) {

                // If the current bar is higher than anything
                // we've previously seen from the left,
                // this becomes our new left boundary.
                if (height[left] >= leftMax) {
                    leftMax = height[left]
                } else {
                    // Otherwise, leftMax is taller than the
                    // current bar, so water can be trapped here.
                    //
                    // For example:
                    //
                    // leftMax = 3
                    // height[left] = 1
                    //
                    // We can store:
                    //
                    // 3 - 1 = 2 units of water.
                    water += leftMax - height[left]
                }

                // Move the left pointer toward the center.
                left++

            } else {

                // We are processing the right side because
                // its current height is smaller.
                //
                // If this bar is higher than anything we've
                // seen from the right, it becomes the new
                // right boundary.
                if (height[right] >= rightMax) {
                    rightMax = height[right]
                } else {
                    // Otherwise, rightMax is taller than the
                    // current bar, so water can be trapped here.
                    water += rightMax - height[right]
                }

                // Move the right pointer toward the center.
                right--
            }
        }

        // Return the total amount of trapped water.
        return water
    }
}